lib/perlfaq4.pod

```diff
@@ -273,32 +273,63 @@ are left as an exercise to the inclined reader.
 
 =head2 Why doesn't & work the way I want it to?
 
-The behavior of binary arithmetic operators depends on whether they're
-used on numbers or strings. The operators treat a string as a series
-of bits and work with that (the string C<"3"> is the bit pattern
-C<00110011>). The operators work with the binary form of a number
-(the number C<3> is treated as the bit pattern C<00000011>).
-
-So, saying C<11 & 3> performs the "and" operation on numbers (yielding
-C<3>). Saying C<"11" & "3"> performs the "and" operation on strings
-(yielding C<"1">).
-
-Most problems with C<&> and C<|> arise because the programmer thinks
-they have a number but really it's a string or vice versa. To avoid this,
-stringify the arguments explicitly (using C<""> or C<qq()>) or convert them
-to numbers explicitly (using C<0+$arg>). The rest arise because
-the programmer says:
-
-    if ("\020\020" & "\101\101") {
-        # ...
-    }
+Perl's C<&> bitwise operator works on both numbers and strings,
+sometimes producing surprising results when you expected a number
+but received a string. You probably expected perl to automatically
+convert the operands to numbers like the mathematical operators would.
+Instead, perl treats string operands as a bitvectors.
 
-but a string consisting of two null bytes (the result of C<"\020\020"
-& "\101\101">) is not a false value in Perl. You need:
+Consider the bitwise difference between the number 3 and the bitvector
+represented by "3". A number has the bit pattern for its magnitude. The
+number 3 is 0b11 (a 2 and a 1). The bitvector has the bit pattern that
+is the ordinal value for each octet, and that value is unrelated to any
+numeric value that the digit represents. The character "3" is the
+bitvector 0b0011_0011.
 
-    if ( ("\020\020" & "\101\101") !~ /[^\000]/) {
-        # ...
-    }
+These operations have different results even though you might think they
+look like the same "number":
+
+	 11  &  3;   # 0b0000_1011 & 0b0000_0011
+	             #     -> 0b0000_0011   (number 3)
+	"11" & "3";  # 0b0011_0001_0011_0001 & 0b0011_0011
+	             #     -> 0b0011_0001   (ASCII char "1")
+
+Note that if any operand has a numeric value, perl uses numeric
+semantics (although you should not count on this):
+
+	my($i, $j) = ( 11,   3 );   # $i & $j  # 11  &  3 -> 3
+ 	my($i, $j) = ("11",  3 );   # $i & $j  # 11  &  3 -> 3
+ 	my($i, $j) = ("11", "3");   # $i & $j  # "11"  &  "3" -> 1
+
+Remember that a perl scalar can have both string and numeric values at
+the same time. A value that starts as a string and has never encountered
+a numeric operation has no numeric value yet. Perl does this to save
+time and work so it doesn't have to decide a numeric value for a scalar
+it might never use as a number. In that case, string sematics wins. But,
```

Comment:
```suggestion
it might never use as a number. In that case, string semantics wins. But,
```

Reply:
fixed in ddd936a77b5b6f4aa41024f68d7b2e1013ecfaec (which isn't showing up yet)